src/FSharp.Data.GraphQL.Server.AspNetCore/Giraffe/HttpHandlers.fs

```diff
@@ -0,0 +1,175 @@
+open Microsoft.Extensions.DependencyInjection
+open Microsoft.Extensions.Logging
+open System.Collections.Generic
+open System.Text.Json
+open System.Text.Json.Serialization
+open System.Threading
+open System.Threading.Tasks
+
+type HttpHandler = HttpFunc -> HttpContext -> HttpFuncResult
+
+module HttpHandlers =
+    open System.Collections.Immutable
+
+    let private httpOk (cancellationToken : CancellationToken) (customHandler : HttpHandler) (serializerOptions : JsonSerializerOptions) payload : HttpHandler =
+        setStatusCode 200
+        >=> customHandler
+        >=> (setHttpHeader "Content-Type" "application/json")
+        >=> (fun _ ctx ->
+                JsonSerializer
+                    .SerializeAsync(
+                        ctx.Response.Body,
+                        payload,
+                        options = serializerOptions,
+                        cancellationToken = cancellationToken
+                    )
+                    .ContinueWith(fun _ -> Some ctx) // what about when serialization fails? Maybe it will never at this stage anyway...
+            )
+
+    let private prepareGenericErrors (errorMessages : string list) =
+        (NameValueLookup.ofList
+            [ "errors",
+                upcast
+                    ( errorMessages
+                        |> List.map
+                            (fun msg ->
+                                NameValueLookup.ofList ["message", upcast msg]
+                            )
+                    )
+            ]
+        )
+
+    /// HttpHandler for handling GraphQL requests with Giraffe.
+    /// This one is for specifying an interceptor when you need to
+    /// do a custom handling on the response. For example, when you want
+    /// to add custom headers.
+    let handleGraphQLWithResponseInterception<'Root>
+      (cancellationToken : CancellationToken)
+      (logger : ILogger)
+      (interceptor : HttpHandler)
+      (next : HttpFunc) (ctx : HttpContext) =
+      task {
+        let cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, ctx.RequestAborted).Token
+        if cancellationToken.IsCancellationRequested then
+            return (fun _ -> None) ctx
+        else
+            let options = ctx.RequestServices.GetRequiredService<GraphQLOptions<'Root>>()
+            let executor = options.SchemaExecutor
+            let rootFactory = options.RootFactory
+            let serializerOptions = options.SerializerOptions
+            let deserializeGraphQLRequest () =
+              task {
+                try
+                  let! deserialized =
+                    JsonSerializer.DeserializeAsync<GraphQLRequest>(
+                      ctx.Request.Body,
+                      serializerOptions
+                    )
+                  return Ok deserialized
+                with
+                | :? GraphQLException as ex ->
+                  return Result.Error [sprintf "%s" (ex.Message)]
```

Comment:
I've made it better in my rebase version of your branch

Reply:
Can you please elaborate on that?

Reply:
I mean the whole handler rewrite with AST parsing, getting JsonSerializerSettings, etc.